index.ts

```diff
@@ -1,165 +1,237 @@
+  }
+  newEventHandler(audioBytesEvent);
+}
+
+let transcriptionMutex = false;
+const transcriptionEventHandler = async (event: AudioBytesEvent) => {
+  // TODO: Unbounded linear growth. Instead, walk backwards or something.
+  const lastCut = getCutTimestamp();
+  const audioBytesEvents = eventlog.events.filter(e => e.eventType === 'audioBytes' && e.timestamp >= lastCut);
+  const joinedBuffer = Buffer.concat(
+    audioBytesEvents.map((event) => event.data.buffer)
   );
-  const result: string = await globalLlamaPromise;
-  mostRecentTranscription.precannedLLMResponse = result;
 
-  const userSpeech: Speech = { speaker: "user", response: userTranscriptionSinceWatermark };
-  conversation.canonicalDialogue.push(userSpeech);
+  // TODO: Wait for 1s, because whisper bindings currently throw out if not enough audio passed in
+  // Therefore fix whisper
+  if (!transcriptionMutex && joinedBuffer.length > ONE_SECOND) {
+    transcriptionMutex = true;
+    globalWhisperPromise = whisper.whisperInferenceOnBytes(joinedBuffer);
+    const transcription = await globalWhisperPromise;
+    const transcriptionEvent: TranscriptionEvent = {
+      timestamp: Number(Date.now()),
+      eventType: 'transcription',
+      data: {
+        buffer: joinedBuffer,
+        transcription,
+        lastAudioByteEventTimestamp: audioBytesEvents[audioBytesEvents.length - 1].timestamp
+      }
+    }
+    newEventHandler(transcriptionEvent);
+    transcriptionMutex = false;
+  }
+}
 
-  const agentSpeech: Speech = { speaker: "agent", response: mostRecentTranscription.precannedLLMResponse || '' };
-  conversation.watermark = mostRecentTranscription.endByte;
-  conversation.canonicalDialogue.push(agentSpeech);
-  console.log('response completed: ');
-  console.log(conversation.canonicalDialogue);
+const cutTranscriptionEventHandler = async (event: TranscriptionEvent) => {
+  const cutTranscriptionEvents = eventlog.events.filter(e => e.eventType === 'cutTranscription');
+  const lastCut = cutTranscriptionEvents.length > 0 ? cutTranscriptionEvents[cutTranscriptionEvents.length - 1].timestamp : eventlog.events[0].timestamp;
+  const timeDiff = event.timestamp - lastCut;
+  if (timeDiff > BUFFER_LENGTH_MS) {
+    const cutTranscriptionEvent: CutTranscriptionEvent = {
+      timestamp: event.timestamp,
+      eventType: 'cutTranscription',
+      data: {
+        buffer: event.data.buffer,
+        transcription: event.data.transcription,
+        lastAudioByteEventTimestamp: event.data.lastAudioByteEventTimestamp
+      }
+    }
+    newEventHandler(cutTranscriptionEvent);
+  }
+}
+
+const responseReflexEventHandler = async (): Promise<void> => {
+  await globalWhisperPromise;
+  const responseReflexEvent: ResponseReflexEvent = {
+    timestamp: Number(Date.now()),
+    eventType: 'responseReflex',
+    data: {
+      transcription: getTransciptionSoFar()
+    }
+  }
+  newEventHandler(responseReflexEvent);
+}
+
+const talkEventHandler = (event: ResponseReflexEvent): void => {
```

Comment:
In DAG, the upstream should be able to pass its state to the downstream tasks.

However here, we are imposing a restriction that only `ResponseReflexEvent` can be processed by this handler and so only `responseReflex` upstream task can effectively call this.

ideally you want something that's a wrapper around this, so can any upstream task can pass its state and then you can simply handle it or discard it in downstream task.

Reply:
Yes, I actually came across this problem. I wanted to have cutTranscriptionEventHandler handle from responseReflex, so I then had to take the generic event class. I think I'm actually using the types wrong probably

I think I might just make the downstream calls implicit rather than explicitly defined in an object that gets looked up, I think it's a bit unclear. And that should solve that problem